c2/c2-client-bundle/c2-client-service/src/main/java/org/apache/nifi/c2/client/service/operation/DebugOperationHandler.java

```diff
@@ -0,0 +1,257 @@
+        return operationAckCreate.apply(operationState);
+    }
+
+    private Function<C2OperationState, C2OperationAck> operationAck(C2Operation operation) {
+        return state -> {
+            C2OperationAck operationAck = new C2OperationAck();
+            operationAck.setOperationId(ofNullable(operation.getIdentifier()).orElse(EMPTY));
+            operationAck.setOperationState(state);
+            return operationAck;
+        };
+    }
+
+    private C2OperationState operationState(OperationState operationState, String details) {
+        C2OperationState state = new C2OperationState();
+        state.setState(operationState);
+        state.setDetails(details);
+        return state;
+    }
+
+    private Stream<Path> debugBundlePaths(String baseDir, DebugBundleFile... debugBundleFiles) {
+        return Stream.of(debugBundleFiles)
+            .map(bundleFile -> Paths.get(baseDir, bundleFile.getFileName()))
+            .filter(Files::exists)
+            .filter(Files::isRegularFile);
+    }
+
+    private List<Path> filterContent(List<Path> bundleFilePaths) {
+        List<Path> contentFilteredFilePaths = new ArrayList<>();
+        for (Path path : bundleFilePaths) {
+            String fileName = path.getFileName().toString();
+            try {
+                Path tempDirectory = createTempDirectory(null);
+                Path tempFile = Paths.get(tempDirectory.toAbsolutePath().toString(), fileName);
+                write(tempFile, (Iterable<String>) lines(path).filter(contentFilter)::iterator);
+                contentFilteredFilePaths.add(tempFile);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+        return contentFilteredFilePaths;
+    }
+
+    public Optional<byte[]> createDebugBundle(List<Path> filePaths) {
+        if (filePaths.size() == 0) {
+            return empty();
+        }
+        ByteArrayOutputStream byteOutputStream = null;
+        try {
+            byteOutputStream = new ByteArrayOutputStream();
+            try (GzipCompressorOutputStream gzipCompressorOutputStream = new GzipCompressorOutputStream(byteOutputStream);
+                 TarArchiveOutputStream tarOutputStream = new TarArchiveOutputStream(gzipCompressorOutputStream)) {
+                for (Path filePath : filePaths) {
+                    TarArchiveEntry tarArchiveEntry = new TarArchiveEntry(filePath.toFile(), filePath.getFileName().toString());
+                    tarOutputStream.putArchiveEntry(tarArchiveEntry);
+                    copy(filePath, tarOutputStream);
+                    tarOutputStream.closeArchiveEntry();
+                }
+                tarOutputStream.finish();
+            }
+        } catch (Exception e) {
+            LOG.error("Error during create compressed bundle", e);
+            closeQuietly(byteOutputStream);
```

Comment:
can't we use try with resources here also?
Also It's not clear why do we need to call the close here and in the finally as well.

Reply:
Unfortunately we can. I need the ByteArrayOutputStream reference to return with byte[] in the end. However if I return with byteOutputStream.toByteArray() from the try-with-resources block, it will return with a malformed result. It seems byteOutputStream needs to be closed before it can be converted to a byte array.
Regarding the two closeQuietly, I had to close the stream in the catch block before nulling the reference out, so it couldn't be closed in the finally clause. I changed this part to be more straightforward.